bug: `CAST(<array>)` causes internal error

### Describe the bug

The SQL `SELECT CAST(MAKE_ARRAY(1, 2, 3)` works just fine. But if you add a cast in there, eg
`SELECT CAST(MAKE_ARRAY(1, 2, 3) AS VARCHAR[])` or `SELECT CAST(MAKE_ARRAY() AS VARCHAR[])`, then I get

```
E       Exception: Optimizer rule 'simplify_expressions' failed
E       caused by
E       simplify_expressions
E       caused by
E       Internal error: Failed due to a difference in schemas, original schema: DFSchema { fields: [DFField { qualifier: None, field: Field { name: "Cast(Array(Array), array<string>)", data_type: List(Field { name: "item", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }, new schema: DFSchema { fields: [DFField { qualifier: None, field: Field { name: "Cast(Array(Array), array<string>)", data_type: List(Field { name: "field", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {}, functional_dependencies: FunctionalDependencies { deps: [] } }.
E       This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

Found in https://github.com/ibis-project/ibis/actions/runs/9042239055/job/24848519442?pr=8666 via ibis and the python bindings. `datafusion.__version__` reveals `36.0.0`


### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

--- a/datafusion/sqllogictest/test_files/cast.slt
+++ b/datafusion/sqllogictest/test_files/cast.slt
@@ -56,3 +56,16 @@ query I
 SELECT 10::bigint unsigned
 ----
 10
+
+# cast array
+query ?
+SELECT CAST(MAKE_ARRAY(1, 2, 3) AS VARCHAR[])
+----
+[1, 2, 3]
+
+
+# cast empty array
+query ?
+SELECT CAST(MAKE_ARRAY() AS VARCHAR[])
+----
+[]

